text/0000-layer-origin-metadata.md

```diff
@@ -0,0 +1,234 @@
+# Motivation
+[motivation]: #motivation
+
+- Layer generation can be influenced by the state of the file system,
+  namely by the state of the application directory, other layers, and
+  the stack's build image
+- Layers aren't reproducible, because we can't know enough about the
+  state of the file system at the instant that the buildpack build
+  executable is invoked (TODO: figure out if there's a less wordy name
+  for this)
+- Builds aren't reproducible, because their constituent layers aren't
+  reproducible
+- Layers and builds should be reproducible
+- Exposing layer origin metadata will enable us to know crucial facts
+  about the state of the file system at the time that the buildpack
+  build executable is invoked
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Terminology
+
+- **Layer origin metadata** is the missing piece of the
+  reproducibility puzzle; it includes the ID of the restored version
+  of the layer (if applicable), modification date, source version
+  info, and a list of the IDs of the layers that were present at layer
+  generation time.
+- **Layer ID** is the layer's unique ID, whether on disk or on the
+  registry.
+
+## Target persona
+
+This feature will be useful for **buildpack authors** (for reproducing
+layers) and **platform operators** (for reproducing builds).
+
+## Example
+
+Consider some buildpack `yabba` and some application `foo-api`. Let's
+also say that the `yabba` buildpack generates a cacheable build layer
+`yabba/build`.
+
+1. I do a build of the `foo-api` as of git SHA aaa000 involving the
+   `yabba` buildpack, without using any cache.
+2. I do a build of the `foo-api` as of git SHA aaa111 involving the
+   `yabba` buildpack, using the cache from the previous build.
+
+I want to account for all of the inputs that were present during the
+second build:
+
+1. the version of the `foo-api` application
+2. the version of the `yabba` buildpack
+3. the version of the stack's build image
+4. the version of the `yabba/build` layer that was restored from the
+   cache
+
+I can account for all but the last; I neither know that layer's
+identity, nor how it was created (the version of the `foo-api`, the
+version of the `yabba` buildpack, whatever layers existed at that
+time). Therefore, I cannot reproduce the second build.
+
+Now consider that I have access to each build's `report.toml`, and
+that the second build's `report.toml` looks something like this:
+
+```toml
+[image]
+tags = ["index.docker.io/image/name:latest", "index.docker.io/image/name:other-tag"]
+digest = "26cca5b0c787"
+previous-image-digest = "97e47fb9e0a6"
+
+[[layers]]
+buildpack = "yabba"
+name = "build"
+on-export = "reused"
+digest = "1742affe03b5"
+modification-date = "2020-07-09 12:41:35-04:00"
+size = 17813
+previous-layer-digest = "c54bba046158"
+
+cached = true
+launch = false
+build = true
+
+[[layers.application-source]]
+type = "git"
+
+[[layers.application-source.version]]
+sha = "26e8c05b151f57161d64b7a5f844bb04caed0ba3"
```

Comment:
In order to provide correct values for fields like this in the case of reused cached layers we will need to store layer origin metadata in the cache. I am not sure we need to describe that mechanism in this RFC but it seems worth pointing out.

Reply:
What concerns me more is how we provide this information for `cache=false` layers. We could still store origin metadata about those layers in the cache (even if we don't store the entire layer) but depending on cache availability there may be cases where we wish to reuse a layer from the previous app image and we don't have cache origin metadata available.

Maybe we should simply avoid setting these fields when the data is unavailable?

Reply:
regarding how cache availability affects builds or not, i'm not sure what the context is. are you talking about the case where the layer is there but its origin metadata, for whatever reason, isn't there? or maybe about the case where we have to retrieve those via two separate requests, the request for the layer succeeding and the request for its origin metadata failing? 

generally, i can imagine some people insisting on having the data about the previous build, for examples when building images bound for prod - maybe such a build is only valid if all origin metadata is present. i can also imagine some people caring more about build availability, not wanting to sacrifice that for lack of traceability. 

in the case of a build that didn't use the cache, these fields would certainly be absent.

Reply:
> are you talking about the case where the layer is there but its origin metadata, for whatever reason, isn't there?

Yes

In some ways the problem I am describing is the inverse of the problem statement from your original RFC. For `cache=false` layers we always have the existing `layer.metadata` but additional layer origin metadata won't exist on the image and needs to come from the cache (or be supplied in some other fashion). However, the cache from the previous build is not guaranteed to be always available (For example if you are using a volume cache and rebuilding an image with `pack` in a different docker daemon. Or if you deleted the cache volume).

The availability depends on the caching strategy. An image cache is always available but less performant than a volume cache. That is the tradeoff. I suspect we can probably just populate these fields when the data is available and let folks pick the caching strategy that meets their availability or performance requirements.